agent/fw_drupal.c

```diff
@@ -778,8 +823,16 @@ void nr_drupal_enable(TSRMLS_D) {
   if (NRINI(drupal_modules)) {
     nr_php_wrap_user_function(NR_PSTR("module_invoke"),
                               nr_drupal_wrap_module_invoke TSRMLS_CC);
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
+     && !defined OVERWRITE_ZEND_EXECUTE_DATA
+    nr_php_wrap_user_function_before_after_clean(NR_PSTR("module_invoke_all"),
+                                                 nr_drupal_wrap_module_invoke_all_before,
+                                                 nr_drupal_wrap_module_invoke_all_after,
+                                                 nr_drupal_wrap_module_invoke_all_after);
```

Comment:
I'd really prefer to see a separate function `nr_drupal_wrap_module_invoke_all_clean` or `nr_drupal_wrap_module_invoke_all_exception_handler`, and if possible extract common code for `nr_drupal_wrap_module_invoke_all_after` and this new function to a helper method which can operate independent of the after/cleanup context. Let's try to follow to tried and tested [SOLID](https://en.wikipedia.org/wiki/SOLID) design principles that make the code maintainable. Having the `_after` function do two things (handle the _after user function call and handle cleanup after exception) smells like a violation of [Single-responsibility principle](https://en.wikipedia.org/wiki/Single-responsibility_principle).

Reply:
extracted core functionality to a helper and added a "clean" function in [79c7ee8](https://github.com/newrelic/newrelic-php-agent/pull/559/commits/79c7ee84f4d33549eb345c94afaea0a58874d33d)